### PR TITLE
fix: add network cleanup for kill and stop cmd

### LINKED ISSUE #2699: The Iptables  of the published port are not cleaned when the container is killed (not removed)
### Description

The Iptables  of the published port are not cleaned when the container is killed (not removed)
So, if another container starts with the same published port, it cannot be accessed correctly.
It may cause:
- Can't access the exposed port from LAN in bridge mode.  https://github.com/containerd/nerdctl/discussions/2142

### Steps to reproduce the issue

1. Choose a random port
```
# PUBLISH_PORT=$((8000 + $RANDOM % 2000))
# echo $PUBLISH_PORT
8905 
```

2. Run the 1st container with  $PUBLISH_PORT, and kill it
```
root@kay201:~# nerdctl run --name ngx1 -d -p $PUBLISH_PORT:80 docker.m.daocloud.io/nginx:alpine
root@kay201:~# nerdctl kill ngx1
root@kay201:~# iptables-save  |grep $PUBLISH_PORT
```
The iptables rule still exists
```
-A CNI-DN-f9fd857ff3d57680c6bc2 -s 10.4.0.0/24 -p tcp -m tcp --dport 8905 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-f9fd857ff3d57680c6bc2 -s 127.0.0.1/32 -p tcp -m tcp --dport 8905 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-f9fd857ff3d57680c6bc2 -p tcp -m tcp --dport 8905 -j DNAT --to-destination 10.4.0.47:80
-A CNI-HOSTPORT-DNAT -p tcp -m comment --comment "dnat name: \"bridge\" id: \"k8s.io-ea6bc8469fbe0c245c9893f14ac5a652d464df7f38932053d2c1cd940c97d7a8\"" -m multiport --dports 8905 -j CNI-DN-f9fd857ff3d57680c6bc2
```

3. Run the 2rd container with  $PUBLISH_PORT, and try to access it.

```
root@kay202:~# nerdctl run --name ngx2 -d -p $PUBLISH_PORT:80 docker.m.daocloud.io/nginx:alpine
root@kay202~# curl 127.0.0.1:$PUBLISH_PORT
curl: (7) Failed to connect to 127.0.0.1 port 8905 after 3075 ms: No route to host
```
**It cannot access**, because There are two port-mapping in the iptables rule

```
root@kay201:~#  iptables-save  |grep $PUBLISH_PORT
-A CNI-DN-eeade65af7cda4d8daf03 -s 10.4.0.0/24 -p tcp -m tcp --dport 8905 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-eeade65af7cda4d8daf03 -s 127.0.0.1/32 -p tcp -m tcp --dport 8905 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-eeade65af7cda4d8daf03 -p tcp -m tcp --dport 8905 -j DNAT --to-destination 10.4.0.48:80
-A CNI-DN-f9fd857ff3d57680c6bc2 -s 10.4.0.0/24 -p tcp -m tcp --dport 8905 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-f9fd857ff3d57680c6bc2 -s 127.0.0.1/32 -p tcp -m tcp --dport 8905 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-f9fd857ff3d57680c6bc2 -p tcp -m tcp --dport 8905 -j DNAT --to-destination 10.4.0.47:80
-A CNI-HOSTPORT-DNAT -p tcp -m comment --comment "dnat name: \"bridge\" id: \"k8s.io-ea6bc8469fbe0c245c9893f14ac5a652d464df7f38932053d2c1cd940c97d7a8\"" -m multiport --dports 8905 -j CNI-DN-f9fd857ff3d57680c6bc2
-A CNI-HOSTPORT-DNAT -p tcp -m comment --comment "dnat name: \"bridge\" id: \"k8s.io-0e68aa46d713a89030af74d16b95e9c44fa30190366e4444031a5d155c1cecb7\"" -m multiport --dports 8905 -j CNI-DN-eeade65af7cda4d8daf03
```

4. Remove the 1st container,  the 2rd container can be accessed.

```
root@kay201:~# nerdctl rm -f ngx1
root@kay201:~# curl 127.0.0.1:$PUBLISH_PORT
OK
```

### Describe the results you received and expected

The Iptables of the published port are not cleaned when the container is killed like docker.

### What version of nerdctl are you using?

root@kay201:~# nerdctl  version
Client:
 Version:       v1.7.0
 OS/Arch:       linux/amd64
 Git commit:    e674fe7ba6e49f12e88cd9c6c442e7ea5232502c
 buildctl:
  Version:      v0.12.3
  GitCommit:    438f47256f0decd64cc96084e22d3357da494c27

Server:
 containerd:
  Version:      v1.7.8
  GitCommit:    8e4b0bde866788eec76735cc77c4720144248fb7
 runc:
  Version:      1.1.10
  GitCommit:    v1.1.10-0-g18a0cb0f

### Are you using a variant of nerdctl? (e.g., Rancher Desktop)

None

### Host information

_No response_

### PR DESCRIPTION
**Pull request by alegrey91**
Fix #2699

This PR provides code to cleanup CNI network setup when `kill` and `stop` command are typed.
Additionally, the `onStartContainer` hook was added in order to restore the CNI network setup when `start` command is typed (as suggested in the issue by @yankay).

## How to test

In order to test this PR, follow these commands:

```
# compile the binary with the latest changes
make

# create the first container with the built nerdctl version
sudo _output/nerdctl run --name ngx1 -d -p 9999:80 nginx:alpine

# check iptables rules have been setup
sudo iptables-save | grep 9999

# kill the first container previosly created
sudo _output/nerdctl kill ngx1

# (now the iptables rules should disappeared
# create the second container bound on the same port
sudo _output/nerdctl run --name ngx2 -d -p 9999:80 nginx:alpine

# check iptables rules have been setup
sudo iptables-save | grep 9999

# access the second container forward
curl 127.0.0.1:9999
```

## Credits

Thanks to @yankay and @89luca89 for the help

### CODE REVIEW

**Comment by alegrey91** [comment]
> Can we have an integration test? (See `cmd/nerdctl/*_test.go`)

Sure!

**Comment by alegrey91** [comment]
> Can we have an integration test? (See `cmd/nerdctl/*_test.go`)

@AkihiroSuda just added

File: cmd/nerdctl/container_stop_linux_test.go [hunk 1]
```diff
@@ -86,3 +88,33 @@ echo "signal quit"`).AssertOK()
...
+		testContainerName = "ngx"
+	)
+	base := testutil.NewBase(t)
+	defer func() {
+		base.Cmd("rm", "-f", testContainerName).Run()
+	}()
+
+	// skip if rootless
+	if rootlessutil.IsRootless() {
+		t.Skip("Nydusify check is not supported rootless mode.")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/container_stop_linux_test.go:104]: How is this relevant to Nydusify?

File: cmd/nerdctl/container_kill_linux_test.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		testContainerName = "ngx"
+	)
+	base := testutil.NewBase(t)
+	defer func() {
+		base.Cmd("rm", "-f", testContainerName).Run()
+	}()
+
+	// skip if rootless
+	if rootlessutil.IsRootless() {
+		t.Skip("Nydusify check is not supported rootless mode.")
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply cmd/nerdctl/container_kill_linux_test.go:46]: sorry, I forgot to change it

File: cmd/nerdctl/container_kill_linux_test.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	containerID := strings.TrimSpace(cmd.Run().Stdout())
+	assert.Equal(t, forwardExists(t, ipt, testutil.Namespace, containerID), true)
+
+	base.Cmd("kill", testContainerName).AssertOK()
+	assert.Equal(t, forwardExists(t, ipt, testutil.Namespace, containerID), false)
+}
+
+// forwardExists check that at least 2 rules are present in the CNI-HOSTPORT-DNAT chain
+// and checks for regex matches in the list of rules
+func forwardExists(t *testing.T, ipt *iptables.IPTables, namespace, containerID string) bool {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/container_kill_linux_test.go:66]: This should be in a new package like `pkg/testutil/iptables`

File: cmd/nerdctl/container_kill_linux_test.go (hunk 2 above)
Comment (AkihiroSuda) [inline cmd/nerdctl/container_kill_linux_test.go:46]:
```suggestion
		t.Skip("pkg/testutil/iptables does not support rootless")
```

File: cmd/nerdctl/container_kill_linux_test.go (hunk 3 above)
Comment (alegrey91) [reply cmd/nerdctl/container_kill_linux_test.go:66]: ok, I got your point, will provide it soon

File: cmd/nerdctl/container_kill_linux_test.go [hunk 4]
```diff
@@ -18,12 +18,12 @@ package main
 
 import (
 	"fmt"
-	"regexp"
 	"strings"
 	"testing"
 
 	"github.com/containerd/nerdctl/v2/pkg/rootlessutil"
 	"github.com/containerd/nerdctl/v2/pkg/testutil"
+	iptablesUtil "github.com/containerd/nerdctl/v2/pkg/testutil/iptables"
```
Comment (AkihiroSuda) [inline cmd/nerdctl/container_kill_linux_test.go:26]:
```suggestion
	iptablesutil "github.com/containerd/nerdctl/v2/pkg/testutil/iptables"
```

Or just call it "iptables"

File: pkg/testutil/iptables/iptablesutil.go
Comment (AkihiroSuda) [inline pkg/testutil/iptables/iptablesutil.go:1]: The file name should correspond to the directory name

Comment (alegrey91) [reply pkg/testutil/iptables/iptablesutil.go:1]: right

File: cmd/nerdctl/container_kill_linux_test.go (hunk 4 above)
Comment (alegrey91) [reply cmd/nerdctl/container_kill_linux_test.go:26]: unfortunately we already have `iptables` used by the package below: `"github.com/coreos/go-iptables/iptables"`.
That's why I was using a different name.

Comment (AkihiroSuda) [reply cmd/nerdctl/container_kill_linux_test.go:26]:
```suggestion
	iptablesutil "github.com/containerd/nerdctl/v2/pkg/testutil/iptables"
```

**Comment by AkihiroSuda** [comment]
Thanks, looks good, but please squash commits

**Comment by alegrey91** [comment]
> Thanks, looks good, but please squash commits

Sure and thanks for the accurate review. I learned something new :)

**Comment by AkihiroSuda** [comment]
CI is failing, probably you need to wrap the clean up function with `WithDetachedNetNSIfAny`
https://github.com/search?q=repo%3Acontainerd%2Fnerdctl%20WithDetachedNetNSIfAny&type=code

**Comment by alegrey91** [comment]
> CI is failing, probably you need to wrap the clean up function with `WithDetachedNetNSIfAny` https://github.com/search?q=repo%3Acontainerd%2Fnerdctl%20WithDetachedNetNSIfAny&type=code

Fixed!
Any idea why the `test-integration-docker-compatibility` test is failing?

**Comment by AkihiroSuda** [comment]
> Any idea why the `test-integration-docker-compatibility` test is failing?

Because the iptables names are different for Docker.

```
=== RUN   TestStopCleanupForwards
    iptables.go:31: error listing rules in chain: "running [/usr/sbin/iptables -t nat -S CNI-HOSTPORT-DNAT --wait]: exit status 1: iptables v1.8.7 (nf_tables): chain `CNI-HOSTPORT-DNAT' in table `nat' is incompatible, use 'nft' tool.\n\n"
    iptables.go:35: not enough rules: 0
    container_stop_linux_test.go:117: assertion failed: false (bool) != true (true bool)
--- FAIL: TestStopCleanupForwards (0.51s)
```

**Comment by alegrey91** [comment]
> > Any idea why the `test-integration-docker-compatibility` test is failing?
> 
> Because the iptables names are different for Docker.
> 
> ```
> === RUN   TestStopCleanupForwards
>     iptables.go:31: error listing rules in chain: "running [/usr/sbin/iptables -t nat -S CNI-HOSTPORT-DNAT --wait]: exit status 1: iptables v1.8.7 (nf_tables): chain `CNI-HOSTPORT-DNAT' in table `nat' is incompatible, use 'nft' tool.\n\n"
>     iptables.go:35: not enough rules: 0
>     container_stop_linux_test.go:117: assertion failed: false (bool) != true (true bool)
> --- FAIL: TestStopCleanupForwards (0.51s)
> ```

What if I edit the `ForwardExists` function to pass the chain as argument?
The new signature will be something like this:
```go
func ForwardExists(t *testing.T, ipt *iptables.IPTables, chain, containerIP string, port int) bool
```
Then we could add a condition in the tests to check if target is docker or not and pass the chain name accordingly.
eg:
```
var chain string
if testutil.GetTarget() == testutil.Docker {
        // docker chain
	chain = "DOCKER"
} else {
        // nerdctl chain
	chain = cniplugin.MustFormatChainNameWithPrefix(netName, containerID, "DN-")
}

assert.Equal(t, iptablesutil.ForwardExists(t, ipt, chain, testutil.Namespace, containerID), true)
```

WDYT?

**Comment by alegrey91** [comment]
@AkihiroSuda I just pushed the code with the changes I described below.
One job of the rootless suite is failing. Not sure if is due to my code or different error.
Can you please rerun it in order to check this thing?
I've added another commit with latest changes. If you think this is good, I can squash the last commit with the previous one.

**Comment by AkihiroSuda** [comment]
Thanks, looks good, please squash

**Review by AkihiroSuda — APPROVED** [review]
Thanks

File: pkg/cmd/container/stop.go [hunk 5]
```diff
@@ -35,6 +35,9 @@ func Stop(ctx context.Context, client *containerd.Client, reqs []string, opt typ
 			if found.MatchCount > 1 {
 				return fmt.Errorf("multiple IDs found with provided prefix: %s", found.Req)
 			}
+			if err := cleanupNetwork(ctx, found.Container, opt.GOptions); err != nil {
+				return fmt.Errorf("unable to cleanup network for container: %s", found.Req)
```
Comment (sebastian-pf9) [inline pkg/cmd/container/stop.go:39]: We're facing this right now. Do you remember, was there any good reason to omit the error message here?

Comment (alegrey91) [reply pkg/cmd/container/stop.go:39]: Hi, as far as I remember, there was no particular reason for that.
You could open a PR to fix it, or I can do it later. Let me know.

Comment (sebastian-pf9) [reply pkg/cmd/container/stop.go:39]: Maybe. Right now this doesn't appear to be annoying enough for me. Thank you!

**Pull request merged**